### PR TITLE
Add shared mode

### PR DESCRIPTION
**Pull request by ageorg29**
Add `shared` mode between `contended` and `uncontended`

### CODE REVIEW

File: ocaml/testsuite/tests/typing-modes/portable-contend.ml [hunk 1]
```diff
@@ -46,17 +46,50 @@ let foo (r @ contended) = {r with b = best_bytes ()}
...
+[%%expect{|
+Line 1, characters 23-24:
+1 | let foo (r @ shared) = r.a <- 42
+                           ^
+Error: This value is shared but expected to be uncontended.
+  Hint: In order to write into the mutable fields,
   this record needs to be uncontended.
 |}]
 
+(* Reading from a mutable field in a shared record is fine *)
```
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-modes/portable-contend.ml:65]: This comment doesn't seem to go with this test -- I don't see reading of a mutable field.

File: ocaml/testsuite/tests/typing-modes/portable-contend.ml [hunk 2]
```diff
@@ -143,11 +184,38 @@ let foo (r @ contended) =
 Line 3, characters 6-16:
 3 |     | [| x; y |] -> ()
           ^^^^^^^^^^
-Error: This value is contended but expected to be uncontended.
+Error: This value is contended but expected to be shared.
   Hint: In order to read from the mutable fields,
-  this record needs to be uncontended.
+  this record needs to be at least shared.
```
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-modes/portable-contend.ml:189]: It's an array, not a record. (This is not a new bug.)

Perhaps include a
```
(* CR modes: Error message should mention array, not record. *)
```

File: ocaml/testsuite/tests/typing-modes/portable-contend.ml [hunk 3]
```diff
@@ -143,11 +184,38 @@ let foo (r @ contended) =
...
+Line 1, characters 33-34:
+1 | let foo (r @ shared) = Array.set r 42 (best_bytes ())
+                                     ^
+Error: This value is shared but expected to be uncontended.
 |}]
 
+(* The signature of Array.get could be generalized to expect shared rather than
+   uncontended, but this would require a change to stdlib. For now the following
+   test fails *)
+let foo (r @ shared) = Array.get r 42
```
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-modes/portable-contend.ml:203]:
```suggestion
(* CR modes: Fix this *)
let foo (r @ shared) = Array.get r 42
```

File: ocaml/testsuite/tests/typing-modes/portable-contend.ml [hunk 4]
```diff
@@ -225,15 +301,33 @@ Error: This function when partially applied returns a value which is nonportable
        but expected to be portable.
 |}]
 
+(* closing over shared gives nonportable *)
+let foo : 'a @ shared portable -> (unit -> unit) @ portable = fun a () -> ()
+[%%expect{|
+Line 1, characters 62-76:
+1 | let foo : 'a @ shared portable -> (unit -> unit) @ portable = fun a () -> ()
```
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-modes/portable-contend.ml:308]: This surprises me. The returned closure does not use the fact that `a` is shared. So (I think) we can submode the `a` to be `contended` and then close over the contended `a` without make the closure `nonportable`.

File: ocaml/testsuite/tests/typing-modes/portable-contend.ml
Comment (goldfirere) [inline ocaml/testsuite/tests/typing-modes/portable-contend.ml:1]: I didn't see a test that closes over a shared value (that is actually read from) and produces a nonportable closure. Could you please add?

**Review by goldfirere — COMMENTED** [review]
Looks good! Just a few tweaks in the tests.

File: ocaml/testsuite/tests/typing-modes/portable-contend.ml (hunk 4 above)
Comment (ageorg29) [reply ocaml/testsuite/tests/typing-modes/portable-contend.ml:322]: Indeed, I see what you mean. This test has the same behavior as: 
`let foo : 'a @ uncontended portable -> (unit -> unit) @ portable = fun a () -> ()`
which also gives an error, even though `a` (here `uncontended`) is not used.

Comment (goldfirere) [reply ocaml/testsuite/tests/typing-modes/portable-contend.ml:322]: OK. We don't have to fix this here. But I think this is another good time to add `(* CR modes: This should be allowed *)` to encourage us to return.

**Review by goldfirere — APPROVED** [review]

**Pull request merged**